### PR TITLE
Executor hardening

### PR DESCRIPTION
**Pull request by jessesuen**
* add retries and memoization for executor k8s API calls
* recover from unexpected panics and annotate the error.

### CODE REVIEW

File: util/retry/retry.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	err = argoerrs.Cause(err)
+
+	if apierr.IsNotFound(err) || apierr.IsForbidden(err) {
+		return true
+	}
+	return false
+}
+
+// IsRetryableNetworkError returns whether or not the error is a retryable network error
+func IsRetryableNetworkError(err error) bool {
```
Comment (jessesuen) [inline util/retry/retry.go]: This is currently unused, and probably needs to be tweaked -- was laying the ground work for retrying artifact load/save errors.

File: workflow/executor/executor.go [hunk 2]
```diff
@@ -315,15 +332,61 @@ func (we *WorkflowExecutor) InitDriver(art wfv1.Artifact) (artifact.ArtifactDriv
...
 // getPod is a wrapper around the pod interface to get the current pod from kube API server
 func (we *WorkflowExecutor) getPod() (*apiv1.Pod, error) {
-	// TODO(jessesuen): harden this as part of issue #653
-	podIf := we.ClientSet.CoreV1().Pods(we.Namespace)
-	pod, err := podIf.Get(we.PodName, metav1.GetOptions{})
+	podsIf := we.ClientSet.CoreV1().Pods(we.Namespace)
+	var pod *apiv1.Pod
+	var err error
+	_ = wait.ExponentialBackoff(DefaultRetry, func() (bool, error) {
+		pod, err = podsIf.Get(we.PodName, metav1.GetOptions{})
```
Comment (sandeepbhojwani) [inline workflow/executor/executor.go]: The underlying client-go request seems to have some built in retry for GET. You may want to check to see if it already meets your retry requirement.

File: util/retry/retry.go (hunk 1 above)
Comment (sandeepbhojwani) [reply util/retry/retry.go]: nice

File: workflow/executor/executor.go (hunk 2 above)
Comment (jessesuen) [reply workflow/executor/executor.go]: Looks like this mechanism does not suit our needs. The issue we are trying to solve is the executor retrying getting pod/secrets from API server to address issues like this:

```
   ├-⚠ hybridize(23)              hm-92j79-3168642571  failed to load artifacts: Get https://10.11.240.1:443/api/v1/namespaces/default/secrets/gcs-credentials: dial tcp 10.11.240.1:443: getsockopt: connection refused
```

**Pull request merged**